docs/proposals/201912_thanos_binary_index_header.md

```diff
@@ -0,0 +1,215 @@
+
+## No Goals
+
+* Removing initial startup for Thanos Store Gateway completely as designed in [Cortex, no initial block sync](https://github.com/thanos-io/thanos/issues/1813)
+  * However this proposal might a step towards that as we might be able to load index-cache/index quickly on demand from disk.
+  * At the same time be able to load `index-header` in query time directly from bucket is not a goal of this proposal.
+* Decreasing size. While it would be nice to use less space; our aim is latency of building/loading the block. That might be correlated with size, but not necessarily (e.g when extra considering compression)
+
+## Verification
+
+* Compare memory and latency usage for startup without index-header being generated on compactor
+* Benchmark for loading the `index-header` vs `index-cache.json` into Store GW.
+
+## Proposal
+
+TSDB index is in binary [format](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md). 
+
+To allow reduced resource consumption and effort when building (1), (2), "index-header" for blocks we plan to reuse similar format for sections like symbols, label indices and 
+posting offsets in separate the file called `index-header` that will replace currently existing `index-cache.json`
+ 
+The process for building this will be as follows:
+
+* Thanks to https://github.com/thanos-io/thanos/pull/1792 we can check final size of index and scan for TOC file.
+* With TOC:
+  * Get the symbols table and copy it to the local file.
+  * Get the posting table and copy it to the local file.
+  * Get the label offset table and copy it to the local file.
+  * Get the label index and copy it to the local file.
+  * Open question: How to get [Posting size to fetch](./201912_thanos_binary_index_header.md#posting-size-to-fetch)?
+  
+With that effort building time and resources should be compared with downloading the prebuilt `index-header` from the bucket. This allows us to reduce complexity of the system as 
+we don't need to cache that in object storage by compactor or even sidecar.
+
+Thanks to this format we can reuse most of the [FileReader](https://github.com/prometheus/prometheus/blob/de0a772b8e7d27dc744810a1a693d97be027049a/tsdb/index/index.go#L664) code
+to load file. TODO: Benchmark if mmap + loading is actually faster than JSON parsing (3)
+
+This also proposes loading blocks on demand (4).
+
+This means maintaining pool with limited number of `index-header` files loaded at the time in Store Gateway.
+With an LRU logic we should be able to deduce which of blocks should be unloaded and left on the disk. 
+
+This proposes following algorithm:
+
+* blocks newer then X (e.g 2w) being always loaded
+* blocks older than X will be loaded on demand on query time and kept cached until evicted.
+* background eviction process unloads blocks without pending reads to the amount of Y (e.g 100 blocks) in LRU fashion. 
+ 
+Both X and Y being configurable.
+ 
+## Risks
+
+### Posting size to fetch
```

Comment:
Added this which we missed initially in this design. Let me know if it's clear. 

TL;DR: we scan `postings` form TSDB index when building `index-cache` and without changing TSDB index this will be still required building `index-header`

@brancz @GiedriusS

Reply:
We got some answers (: